packages/babel-parser/src/parser/expression.js

```diff
@@ -1164,6 +1182,13 @@ export default class ExpressionParser extends LValParser {
       case tt._class:
         node = this.startNode();
         this.takeDecorators(node);
+        if (this.lookahead().type === tt.dot) {
```

Comment:
Agree. I adjusted the way of judgment, the entry of expression, and the consumption time of token.

1. I think the timing of consuming  tokens is independent for each fucntion similar to 'parse...', and other functions should not care, so there is no token consumption for `parseClass` in `parseClassOrClassHasInstanceExpression` .

2. `this.takeDecorators(node);` is executed before parseClass and inside the function, I think that it is not necessary to execute twice, so I made changes in `parseClassOrClassHasInstanceExpression`,

Reply:
You can mark `hasInstance` as a keyword-like token, like we did for `from`, `get`, ... `meta`, any keyword-like identifier that may bear different role than a literal.

https://github.com/babel/babel/blob/29dff68cbeeeb5dcf35586203f730325fd676998/packages/babel-parser/src/tokenizer/types.js#L283-L286

The tokenizer will create different token types for every one of them in `readWord`, as long as they are defined in `types`:

https://github.com/babel/babel/blob/29dff68cbeeeb5dcf35586203f730325fd676998/packages/babel-parser/src/tokenizer/index.js#L1723

We don't have to compare with `"hasInstance"` because for every string V8 will compute the hash and searches in the SwissTable, which backs the JS Map `keywordTypes `.